File: pyforms/web/Controls/ControlBase.py
import uuid

class ControlBase(object):

    _value = None
    _label = None
    _controlHTML = ""

    def __init__(self, label = "", defaultValue = "", helptext=''):
        self._name      = ""
        self._help      = helptext
        self._value     = defaultValue
        self._parent    = None
        self._label     = label
        self._visible   = True
        self._popupMenu = None
        self._id = uuid.uuid4()
        self._controlHTML = ""


    def initControl(self):
        self._controlHTML = "<div id='id{0}' ><input type='text' id='{1}' /></div>".format( self._id, self._name )
        return self._controlHTML

    def serialize(self):
        return { 
            'name':     self.__class__.__name__, 
            'value':    str(self.value),
            'label':    self._label if self._label else '',
            'help':     self._help if self._help else '',
            'visible':  int(self._visible)
        }

    def deserialize(self, properties):
        self.value    = properties.get('value',None)
        self._label   = properties.get('label','')
        self._help    = properties.get('help','')
        self._visible = properties.get('visible',True)

    def finishEditing(self): self.updateControl()

    def updateControl(self): pass

    def changed(self): pass

    def load(self, data):
        if 'value' in data: self.value = data['value']

    def save(self, data):
        if self.value: data['value'] = self.value

    def valueUpdated(self, value): pass

    def show(self): pass

    def hide(self): pass

    def openPopupMenu(self, position): pass

    def addPopupSubMenuOption(self, label, options): pass

    def addPopupMenuOption(self, label, functionAction = None): pass

    def __repr__(self): return self.value

    ############################################################################
    ############ Properties ####################################################
    ############################################################################
    
    @property
    def help(self): return self._help.replace('\n', '&#013;') if self._help else ''

    @property
    def enabled(self): pass

    @enabled.setter
    def enabled(self, value): pass

    ############################################################################

    @property
    def value(self): return self._value

    @value.setter
    def value(self, value):
        oldvalue = self._value
        self._value = value
        if oldvalue!=value: self.valueUpdated(value)

    ############################################################################

    @property
    def label(self): return self._label

    @label.setter
    def label(self, value): self._label = value

    ############################################################################

    @property
    def parent(self): return self._parent

    @parent.setter
    def parent(self, value): self._parent = value



    @property
    def maxWidth(self): return -1

    @maxWidth.setter
    def maxWidth(self, value): pass

    @property
    def minWidth(self): return -1

    @minWidth.setter
    def minWidth(self, value): pass


    @property
    def maxHeight(self): return -1

    @maxHeight.setter
    def maxHeight(self, value): pass

    @property
    def minHeight(self): return -1

    @minHeight.setter
    def minHeight(self, value): pass

    def __str__(self): return "<span id='place-{0}-{1}' />".format(self.parent._id, self._name)



    #### Variable connected to the Storage manager of the corrent user
    @property
    def storage(self): return self._storage

    @storage.setter
    def storage(self, value): self._storage = value
    #######################################################

    #### This variable has the current http request #######
    @property
    def httpRequest(self): return self._httpRequest

    @httpRequest.setter
    def httpRequest(self, value): self._httpRequest = value
    #######################################################

File: pyforms/web/django/static/pyformsjs/ControlBase.js


function ControlBase(name, properties){
	var self = this;

	this.name 			= name;
	this.properties 	= properties;
	this.basewidget 	= undefined; //Will be set in runtime by the parent BaseWidget object.
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.control_id = function(){ 
	return this.basewidget.control_id(this.name); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.jquery = function(){ 
	return $("#"+this.control_id()); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.place_id = function(){ 
	return "place-"+this.control_id(); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.jquery_place = function(){ 
	return $( "#"+this.place_id() ); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.get_value = function(){ 
	return this.jquery().val(); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.set_value = function(value){
	this.jquery().val(value); 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.deserialize = function(data){
	$.extend(this.properties, data);
	this.set_value(this.properties.value);
	if(this.properties.visible) 
		this.jquery_place().show();
	else 
		this.jquery_place().hide();
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.serialize = function(){
	this.properties.value = this.get_value();
	return this.properties; 
};

////////////////////////////////////////////////////////////////////////////////

ControlBase.prototype.init_control = function(){
	if(this.properties.visible) 
		this.jquery_place().show();
	else 
		this.jquery_place().hide();
};